sw-emulator/lib/periph/src/ml_dsa87.rs

```diff
@@ -0,0 +1,1254 @@
+                    .reg
+                    .modify(Status::READY::CLEAR + Status::VALID::CLEAR);
+
+                self.op_complete_action = Some(self.timer.schedule_poll_in(ML_DSA87_OP_TICKS));
+            }
+            _ => {}
+        }
+
+        if self.control.reg.is_set(Control::ZEROIZE) {
+            self.zeroize();
+        }
+
+        Ok(())
+    }
+
+    fn gen_key(&mut self) {
+        let seed_bytes = &bytes_from_words_le(&self.seed);
+        let mut rng = StdRng::from_seed(*seed_bytes);
+        let (pk, sk) = try_keygen_with_rng(&mut rng).unwrap();
+
+        self.pk = words_from_bytes_le(&pk.into_bytes());
+        self.sk_out = words_from_bytes_le(&sk.into_bytes());
+    }
+
+    fn sign(&mut self) {
+        let seed_bytes = &bytes_from_words_le(&self.seed);
+        let mut rng = StdRng::from_seed(*seed_bytes);
+
+        let secret_key_bytes = &bytes_from_words_le(&self.sk_in);
+        let secret_key = PrivateKey::try_from_bytes(*secret_key_bytes).unwrap();
+
+        let message = &bytes_from_words_le(&self.message);
+
+        // The Ml_Dsa87 5 signature is 4595 len but the reg is one byte longer
+        let signature = secret_key.try_sign_with_rng(&mut rng, message).unwrap();
+        let signature_extended = {
+            let mut sig = [0; SIG_LEN + 1];
+            sig[..SIG_LEN].copy_from_slice(&signature);
+            sig
+        };
+        self.signature
+            .copy_from_slice(&words_from_bytes_le(&signature_extended));
+    }
+
+    fn verify(&mut self) {
+        let message = &bytes_from_words_le(&self.message);
+
+        let public_key = {
+            let key_bytes = &bytes_from_words_le(&self.pk);
+            PublicKey::try_from_bytes(*key_bytes).unwrap()
+        };
+
+        let signature = &bytes_from_words_le(&self.signature);
+
+        let result = public_key.verify(message, &signature[..SIG_LEN].try_into().unwrap());
+
+        self.verification_result
+            .iter_mut()
+            .for_each(|e| *e = if result { 1 } else { 0 });
+    }
+
+    fn op_complete(&mut self) {
+        match self.control.reg.read_as_enum(Control::CTRL) {
+            Some(Control::CTRL::Value::GEN_KEY) => self.gen_key(),
+            Some(Control::CTRL::Value::SIGN) => self.sign(),
+            Some(Control::CTRL::Value::VERIFY) => self.verify(),
+            _ => {}
+        }
+
+        self.status
+            .reg
+            .modify(Status::READY::SET + Status::VALID::SET);
+    }
+
+    /// Called by Bus::poll() to indicate that time has passed
+    fn poll(&mut self) {
+        if self.timer.fired(&mut self.op_complete_action) {
+            self.op_complete();
+        }
+    }
+
+    /// Called by Bus::warm_reset() to indicate a warm reset
+    fn warm_reset(&mut self) {
+        // TODO: Reset registers
+    }
+
+    /// Called by Bus::update_reset() to indicate an update reset
+    fn update_reset(&mut self) {
+        // TODO: Reset registers
+    }
+}
+
+#[cfg(test)]
```

Comment:
Due to the length of the tests, I'd suggest one of:

* moving them to a separate file,
* generating the keys from a seed to reduce the size, or
* moving the keys to a separate file and using `include!`.

Reply:
I like the idea of putting kat data in a separate file and using `include!`.